### PR TITLE
Refine bibblio text

### LINKED ISSUE #126: Refine Bibblio submissions
We uploaded our first batch of ~200 english nonfiction works to Bibblio, and there's a good start of recommendations tied especially by author. Because we're sending the full text and it's a text-based recommendation system, though, sometimes there is high term affinity for words like "man" and "thou".

I'd like to mitigate this on our end by refining the text that we send. @leonardr recommended tf-idf documents, and there may also be use for a blacklist.

### PR DESCRIPTION
**Pull request by courte**
This can wait until after multi-tenancy. The branch includes changes I've made to the Bibblio export process--mainly an attempt to strip publisher matter from the text before submission to reduce false recommendations. There have also been a couple small changes to the Bibblio API that are incorporated here, and I tried to reduce duplicates (more on that below).

Fixes #126.

### CODE REVIEW

File: bibblio.py [hunk 1]
```diff
@@ -213,6 +327,21 @@ def output_source(self):
         return DataSource.lookup(self._db, self.DATA_SOURCE_NAME)
 
     def items_that_need_coverage(self, identifiers=None, **kwargs):
+        equivalent = aliased(Identifier)
```
Comment (courte) [inline bibblio.py]: **tl;dr: I'd appreciate thoughts on reducing duplicates from the content db.**
This work with equivalencies (in addition to the removal of suppressed/superceded LicensePools) has all been added to try to reduce sending duplicate content items or Works to Bibblio. I don't have a lot of information about the duplicates through Bibblio, so I'm not sure whether they're different Works that represent the same book or different Editions/Representations that are the same Work.

Basically I don't know if this solution has any merit while adding an additional _db query to an already slow script. I think it might actually be better to do this equivalency comparison at the point that Representations are being selected, but that would make the whole thing take even *longer*, and I don't think it's worth it.

File: bibblio.py [hunk 2]
```diff
@@ -168,6 +167,116 @@ def create_content_item(self, content_item):
...
+        if isinstance(identifier, Identifier):
+            if not identifier.type == Identifier.BIBBLIO_CONTENT_ITEM_ID:
+                raise TypeError('Identifier is not a Bibblio Content Item')
+
+            content_item_id = identifier.identifier
+        delete_url = self.CONTENT_ITEMS_ENDPOINT + content_item_id
+        response = HTTP.request_with_timeout(
+            'DELETE', delete_url, headers=self.default_headers,
+            allowed_response_codes=[200],
+            disallowed_response_codes=['4xx']
```
Comment (leonardr) [inline bibblio.py]: You don't need to specify `disallowed_response_codes` since the only allowed response code is 200.

File: bibblio.py [hunk 3]
```diff
@@ -168,6 +167,116 @@ def create_content_item(self, content_item):
...
+
+    # FILTERED_PHRASES lists strings that are directly related to the
+    # distributor in decreasing order of specificity.
+    FILTERED_PHRASES = None
+
+    @classmethod
+    def filter_spine_idrefs(cls, spine_idrefs):
+        for s in spine_idrefs:
+            if s in cls.SPINE_IDREFS:
+                print "'%s' in SPINE_IDREFS" % s
```
Comment (leonardr) [inline bibblio.py]: This test code should either be removed or converted to use the log system.

File: bibblio.py [hunk 4]
```diff
@@ -168,6 +167,116 @@ def create_content_item(self, content_item):
...
+        '(The)? Project Gutenberg Ebook (of)?',
+        '(The)? Project Gutenberg E(-)?text (of)?',
+        'Project Gutenberg License',
+        'Project Gutenberg',
+        'Gutenberg',
+    ]
+
+    # DISQUALIFYING_PHRASES = [
+    #     'START : FULL (_)? LICENSE (\.|\*|\n)* THE FULL PROJECT GUTENBERG LICENSE'
+    # ]
```
Comment (leonardr) [inline bibblio.py]: What was this for? Does the license have its own spine item and you were using the text of the spine item to disqualify the whole thing?

File: bibblio.py [hunk 5]
```diff
@@ -228,14 +357,27 @@ def items_that_need_coverage(self, identifiers=None, **kwargs):
...
+                    or_(
+                        CustomList.id==self.custom_list.id,
+                        edition_list.id==self.custom_list.id
+                    ),
+                    not_(
+                        or_(
+                            LicensePool.suppressed==True,
+                            LicensePool.superceded==True
+                        )
+                    ),
```
Comment (leonardr) [inline bibblio.py]: Not a big deal but this would be a little simpler if you filtered on suppressed==False and superceded==False. You might also try using `Library.restrict_to_ready_deliverable_works`, which is newly available in core.

File: bibblio.py [hunk 6]
```diff
@@ -312,7 +463,12 @@ def get_full_text(self, work):
             .join(LicensePoolDeliveryMechanism.delivery_mechanism)\
             .filter(
                 LicensePool.work_id==work.id,
-                DeliveryMechanism.drm_scheme==DeliveryMechanism.NO_DRM)\
+                DeliveryMechanism.drm_scheme==DeliveryMechanism.NO_DRM,
+                or_(
+                    Representation.status_code==200,
+                    Representation.status_code==None
+                )
+            )\
```
Comment (leonardr) [inline bibblio.py]: This is the kind of thing you might be able to bypass with `restrict_to_ready_deliverable_works`.

File: bibblio.py (hunk 4 above)
Comment (courte) [inline bibblio.py]: Ugh, this is not supposed to be in the commit. Some Gutenberg books have a file that just has license information in it and nothing else. It doesn't have a consistent name, but it has some consistent text.

Unfortunately, some of them just shove the license into an existing file that is the only file and has the whole book in it. So this was sending empty files up to API and got cut.

File: bin/util/bibblio_export [hunk 7]
```diff
@@ -25,12 +26,24 @@ parser.add_argument(
     '--languages', type=unicode, nargs='*', default=u'eng',
     help='Look for Works in these languages'
 )
+parser.add_argument(
+    '--fiction', action='store_true',
+    help='Export fiction and nonfiction texts'
+)
```
Comment (leonardr) [inline bin/util/bibblio_export]: My understanding is that you normally only include nonfiction, and this _expands_ it to include fiction. But --fiction sounds like you're doing fiction _instead. Calling it --include-fiction would be more clear IMO.

Comment (courte) [inline bin/util/bibblio_export]: That's how it used to be, but I decided (at least for now) to keep fiction and nonfiction separate. Open to thoughts on this; I thought it could help recommendation results.

File: bibblio.py (hunk 6 above)
Comment (courte) [inline bibblio.py]: I tried this and ran into errors with the `exists()` clause in that method not auto-correlated with the required classes -- possibly because this queries Representation instead of Work? Or because LicensePool and LicensePoolDeliveryMechanism had been joined in unexpected ways? I'm not sure, but it didn't seem worth it to keep fighting with it.

File: bin/util/bibblio_export (hunk 7 above)
Comment (leonardr) [inline bin/util/bibblio_export]: I think keeping fiction and nonfiction separate is appropriate. If that's what you're doing then the `help` should be changed to "Export fiction texts instead of nonfiction".

**Comment by leonardr** [comment]
Good to go apart from the `help` thing.
